samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -267,20 +273,41 @@ public TestRunner addOutputStream(CollectionStream stream) {
     return this;
   }
 
+
   /**
    * Utility to run a test configured using TestRunner
+   *
+   * @param timeout in ms for the Test to Run
+   * @throws InterruptedException if the current thread was interrupted
+   * @throws ExecutionException if the samza job threw an exception
+   * @throws TimeoutException if the wait timed out while running Samza job
    */
-  public void run() {
+  public void run(int timeout) throws InterruptedException, ExecutionException, TimeoutException {
     Preconditions.checkState((app == null && taskClass != null) || (app != null && taskClass == null),
         "TestRunner should run for Low Level Task api or High Level Application Api");
     final LocalApplicationRunner runner = new LocalApplicationRunner(new MapConfig(configs));
-    if (app == null) {
-      runner.runTask();
-      runner.waitForFinish();
-    } else {
-      runner.run(app);
-      runner.waitForFinish();
+
+    final Runnable runSamzaJob = new Thread() {
+      @Override
+      public void run() {
+        if (app == null) {
+          runner.runTask();
+          runner.waitForFinish();
+        } else {
+          runner.run(app);
+          runner.waitForFinish();
+        }
+      }
+    };
+
+    final ExecutorService executor = Executors.newSingleThreadExecutor();
```

Comment:
Two points here: 
1. `LocalApplicationRunner` has a bounded `waitForFinish` api which already solves the timeout problem (that is solved by spawning a thread via `ExectorService`).
2. All the threads owned by LocalApplicationRunner api are of deamon type. Ideally any connector(kafka, brooklin etc) should not own/manage non-daemon thread.  So I think it will be unnecessary to forcefully shutdown the `LocalApplicationRunner` threads through `executorService.shutdownNow()`. 

So I think this executor thread block can be replaced with the following. 

```java
        if (app == null) {
          runner.runTask();
          status = runner.waitForFinish(timeout);
        } else {
          runner.run(app);
          status = runner.waitForFinish(timeout);
        }
        ApplicationStatus status = runner.status(app);

	// Add checks here for the application status and throw up to junit if necessary. 
        if (status.getStatusCode() == UnsuccessfulFinish) {
        	throw new SamzaException(status.getThrowable());
        }
```

Currently the `localApplicationRunner.status(app)` api is semantically broken and returns the ApplicationStatus for both the StreamApp, StreamTask.

To get status of StreamTask we have to use `runner.getStatus(app)`

I think we can use that here till we get in Yi's high level API change which will solve this problem(`StreamTaskApplication`, `StreamApplication` & `ApplicationRunnable`).

Reply:
Thanks for this suggestion, will use this and remove ExecutorService, also add more failure tests